build/toolchain/wasm/BUILD.gn

```diff
@@ -29,7 +29,62 @@ gcc_toolchain("wasm") {
   toolchain_cpu = "wasm"
   toolchain_os = "wasm"
 
-  is_clang = true
-
   link_outputs = [ "{{root_out_dir}}/{{target_output_name}}.wasm" ]
+
+  toolchain_args = {
+    is_clang = true
+    target_os = target_os
+    target_cpu = target_cpu
+    current_cpu = toolchain_cpu
+    current_os = toolchain_os
+    skia_use_angle = false
+    skia_use_dng_sdk = false
+    skia_use_dawn = false
+    skia_use_webgl = true
+    skia_use_webgpu = false
+    skia_use_expat = false
+    skia_use_fontconfig = false
+    skia_use_freetype = true
+    skia_use_libheif = false
+    skia_use_libjpeg_turbo_decode = true
+    skia_use_libjpeg_turbo_encode = false
+    skia_use_libpng_decode = true
+    skia_use_libpng_encode = true
+    skia_use_libwebp_decode = true
+    skia_use_libwebp_encode = false
+    skia_use_lua = false
+    skia_use_piex = false
+    skia_use_vulkan = false
+    skia_use_wuffs = true
+    skia_use_zlib = true
+    skia_enable_gpu = true
+    skia_build_for_debugger = false
+    skia_enable_sksl_tracing = false
+    skia_use_icu = true
+    skia_use_harfbuzz = true
+    skia_enable_fontmgr_custom_directory = false
+    skia_enable_fontmgr_custom_embedded = true
+    skia_enable_fontmgr_custom_empty = false
+    skia_enable_skshaper = true
+    skia_enable_skparagraph = true
+    skia_enable_pdf = false
+    skia_canvaskit_force_tracing = false
+    skia_canvaskit_enable_skp_serialization = true
+    skia_canvaskit_enable_effects_deserialization = false
+    skia_canvaskit_enable_skottie = false
+    skia_canvaskit_include_viewer = false
+    skia_canvaskit_enable_particles = false
+    skia_canvaskit_enable_pathops= true
+    skia_canvaskit_enable_rt_shader = true
+    skia_canvaskit_enable_matrix_helper = false
+    skia_canvaskit_enable_canvas_bindings = false
+    skia_canvaskit_enable_font = true
+    skia_canvaskit_enable_embedded_font = true
+    skia_canvaskit_enable_alias_font = true
+    skia_canvaskit_legacy_draw_vertices_blend_mode = false
+    skia_canvaskit_enable_debugger = false
+    skia_canvaskit_enable_paragraph = true
+    skia_canvaskit_enable_webgl = true
+    skia_canvaskit_enable_webgpu = false
```

Comment:
We've typically handled these arguments as things specified in `args.gn` set up by the `gn` utility script in the engine repo. Why do we want to move them here? This seems to further complicate buildroot when we're trying to get rid of it/merge it with the engine repo.

Reply:
The reason I'm doing it this way is because these skia arguments interfere with the Skia build used by Flutter's mobile engines. An example is enabling `skia_enable_webgl` causes Skia to build with the webgl backend, which is wrong for Flutter mobile.

Reply:
Sure, but does that just mean we should build canvaskit in its own `out` directory with its own args?

I think it'll be confusing to see these values set differently in `args.gn` but actually get overridden here.

Reply:
We want to build CanvasKit as part of the `flutter_web_sdk` folder as part of the standard Flutter engine build (e.g. under `host_debug_unopt`)

Reply:
> We want to build CanvasKit as part of the `flutter_web_sdk` folder as part of the standard Flutter engine build (e.g. under `host_debug_unopt`)

@chinmaygarde @zanderso WDYT?

I'm concerned it will be confusing to start introducing artifacts with very different targets/archiectures/build arguments under the same out directory. I'd expect, for example, that we build canvaskit as something like `out/wasm_debug` or something. There are a number of advantages to keeping wasm separate from the `host*` builds too, since it doesn't really differ based on host or target architecture type (so canvaskit.wasm and .js are the same whether you're building on or for x64 or arm64, etc.).

Reply:
To give some extra context on why I'm building CanvasKit as part of the `host` build rather than in its own output folder (e.g. `wasm_debug`):

I used to do it that way (building in its own output folder) but it was a bit ugly because (1) I had to have a lot of if-else statements in `tools/gn` which would disable setting arguments in the case of running a WASM build. This was due to those arguments not being used for building CanvasKit. (2) I had to modify the default top-level target to build the Flutter SDK in the normal case, and only build CanvasKit in the WASM case. This is because if certain GN files are even reachable by import when building for WASM (e.g. the Dart SDK BUILD.gn file) it will cause the build to fail (e.g. because Dart doesn't recognize "wasm" as a valid target cpu).

Having a separate output folder for the CanvasKit build was also causing me to write substantial changes to LUCI and flutter_tool since they would have to build a separate output and pass it along for tests and for the Flutter Web SDK.

Reply:
If we think of web as being a different platform, like iOS and Android are different platforms, then giving it its own `out/web_debug` directory makes sense. I don't know the full catalogue of complications that would introduce, though. It would require a careful dance between the GN build and the recipes, at least. Might be worth doing anyway.

Setting that aside, the way we have handled similar situations in the past is by introducing new targets that hardcode the desired configurations. Like here https://github.com/dart-lang/sdk/blob/main/runtime/configs.gni. Then, you depend on the target with hardcoded config with the toolchain specified (`//foo/bar_with_my_config($my_toolchain)`). That way, target/dependency specific configuration doesn't leak into the toolchain definition, which does seem like a layering violation.

There are way more args here than what exists in the Dart repo, though, and I don't know if we can introduce new targets with these configs specified into the Skia build, so the approach that we've followed in the past might not scale to this situation. Is setting all of these args really necessary? Are some of these being set to the default values?

OTOH, if we can't do the above either because there are too many args, or because we can't introduce new targets, we still might improve on the situation by doing something simple like giving the toolchain a more descriptive name, like `"wasm_for_canvaskit"`, and leaving comments like, "Only use this toolchain to build canvaskit". It might also be possible to write some GN `assert`s to enforce that.

Reply:
I changed it to be called "wasm_for_canvaskit" and added a comment. I'm still not super familiar with configs but I'm unsure if they allow you to override gn args. I'm overriding them in the toolchain because they differ between the Skia build for Flutter and the Skia build for CanvasKit. I am looking into if I can narrow down the list of args overrides.

Reply:
I think global toolchain configs were a big bug in our system, and this PR is a step in the right direction. A single invocation of `gn` + `ninja` should be able to build multiple things targeting different architectures and configurations. Otherwise, what you end up is multiple top-level targets that depend on each other, e.g. a host build of the Dart SDK depending on the WASM build of CanvasKit. So now you need yet another build system on top of this to build all the right targets and stitch them together. On top of that there's the issue that @hterkelsen outlines above, where global configs start pushing each other, with web-specific config leaking into mobile versions. I think we should aim for `gn` to be our only build system that's capable of expressing build graphs spanning target architectures.

It also looks like it's appropriate that we put WASM-specific parameters here, because as of right now this is where we configure architecture-specific toolchains. However, I'm a big fan of avoiding maintaining code across repos, so if there's a graceful way to specify these args inside the engine repo, that may be beneficial. If that refactoring requires a lot of time though, I'd prefer we start with this, unblock others waiting on this change (e.g. Skwasm needs this), and then undertake the refactor.

Would that work?

Reply:
It's not clear to me how this will co-exist properly in a single out directory - for example, what will happen when multiple targets want to build the same Skia TUs with different build args?

Reply:
> I think global toolchain configs were a big bug in our system, and this PR is a step in the right direction. A single invocation of `gn` + `ninja` should be able to build multiple things targeting different architectures and configurations. 

This is possible in GN, but I think it would at least require explicitly plumbing a toolchain in many places where it's currently implicit. I wouldn't be surprised if the additional plumbing needed to cross repo boundaries.

> It's not clear to me how this will co-exist properly in a single out directory - for example, what will happen when multiple targets want to build the same Skia TUs with different build args?

Not sure I'm understanding the concern, but specifying a different toolchain makes a different target. Build results for the default toolchain go directly under `out/`. Build results for a different toolchain go under a subdirectory of `out/`, like `out/clang_x64`.

Reply:
Ahh, right, ok

Reply:
@hterkelsen - do you think you could try what Zach is suggesting?